### DISCUSSION TITLE
How to list all layers manually added to leaflet map by draw_control panel?

### DISCUSSION THREAD

**Discussion by martinbaerwolff** [zauberzeug/nicegui#4028] [Q&A] [answered]
### Question

I am trying to list all layers that have been created manually to a leaflet map using the control panel (to read their coordinates).

Is this possible?

`m.layers` doesn´t include those layers, just generic layers or tile layers.

I also tried `eachLayer` with `m.run_map_method` but couldn´t get the code to run. I tried the following in an async def:

```python
layers = await m.run_map_method("eachLayer", "function(layer){layers.push(layer);}")
```

I ran into this error: `TimeoutError: JavaScript did not respond within 1.0 s`

Is it possible to receive return values from `run_map_method`?

Can be reproduced with the folllowing code:

```python
from nicegui import events, ui

def handle_draw(e: events.GenericEventArguments):
    layer_type = e.args['layerType']
    coords = e.args['layer'].get('_latlng') or e.args['layer'].get('_latlngs')
    ui.notify(f'Drawn a {layer_type} at {coords}')
    
    # Here comes the crucial line:
    layers = await m.run_map_method("eachLayer", "function(layer){layers.push(layer);}")
    ui.notify(layers)

draw_control = {
    'draw': {
        'polygon': True,
        'marker': True,
        'circle': True,
        'rectangle': True,
        'polyline': True,
        'circlemarker': True,
    },
    'edit': {
        'edit': True,
        'remove': True,
    },
}
m = ui.leaflet(center=(51.505, -0.09), draw_control=draw_control)
m.classes('h-96')
m.on('draw:created', handle_draw)
m.on('draw:edited', lambda: ui.notify('Edit completed'))
m.on('draw:deleted', lambda: ui.notify('Delete completed'))

ui.run()
```

Can anyone help?

**Comment by falkoschindler** [comment, marked as answer]
Hi @martinbaerwolff,

I don't think calling `run_map_method("eachLayer", ...)` is the right approach, because this calls a function for each layer separately, so you don't get an answer with all layers.

Apart from that you would need to add a colon like this `run_map_method(":eachLayer", ...)` to indicate that the string argument needs to be converted to a JavaScript function on the client. But we need a different approach anyway.

I'd suggest to write a JavaScript command collecting all layers via `eachLayer`:

```py
@ui.page('/')
def index():
    async def handle_draw(e: events.GenericEventArguments):
        layers = await ui.run_javascript(f'''
            const layers = [];
            getElement({e.sender.id}).map.eachLayer((layer) => {{
                layers.push({{
                    'leaflet_id': layer._leaflet_id,
                    'latlng': layer._latlng,
                    'latlngs': layer._latlngs,
                    'url': layer._url,
                }});
            }});
            return layers;
        ''')
        print(layers)
    ui.leaflet((51.505, -0.09), draw_control={'draw': {}}).on('draw:created', handle_draw)
```

Note that you need to extract individual attribute like "leaflet_id", "latlng" etc. because serializing and sending the whole layer object doesn't work due to circular references. This is what I tried to deal with in PR #3586, but failed to achieve in a general sense.

**Reply by martinbaerwolff** [reply]
Thanks for the quick response @falkoschindler, this works like a charm! 

I did also try `ui.run_javascript()`, but since I am a JS newbie I did not even know that I have to find the right element - `events.GenericEventArguments.sender.id` it is! This also opens so many other options for me. Thanks a lot!

P.S.: Your really built such a nice GUI library for Python people, best I came across yet! Great work!
